build/l10n-plugin.mts

```diff
@@ -0,0 +1,123 @@
+import { Plugin } from 'vite'
+import { loadTranslations } from './translations.mts'
+import { dirname, resolve } from 'path'
+
+/**
+ * This is a plugin to split all translations into chunks of users meaning components that use that translation
+ * If a file imports `t` or `n` from 'l10n.js' that import will be replaced with a wrapper that registeres only the required translations for the file that imports the functions.
+ * Allowing vite to treeshake all not needed translations when building applications
+ *
+ * @param dir Path to the l10n directory for loading the translations
+ */
+export default (dir: string) => {
+	// mapping from filesnames -> variable name
+	let nameMap: Record<string, string>
+	// all loaded translations, as filenames ->
+	const translations: Record<string, { l: string, t: Record<string, { v: string[], p?: string }> }[]> = {}
+
+	return {
+		name: 'nextcloud-l10n-plugin',
+		enforce: 'pre',
+
+		/**
+		 * Prepare l10n loading once the building start, this loads all translations and splits them into chunks by their usage in the components.
+		 */
+		async buildStart() {
+			this.info('[l10n] Loading translations')
+			// all translations for all languages and components
+			const allTranslations = await loadTranslations(dir)
+
+			this.info('[l10n] Loading translation mapping for components')
+			// mapping which files (filename:filename2:filename3) contain which message ids
+			const context = (await import('./extract-l10n.mjs')).context
+			nameMap = Object.fromEntries(Object.keys(context).map((key, index) => [key, `t${index}`]))
+
+			this.info('[l10n] Building translation chunks for components')
+			// This will split translations in a map like "using file(s)" => {locale, translations}
+			for (const locale in allTranslations) {
+				const currentTranslations = allTranslations[locale]
+				for (const [usage, msgIds] of Object.entries(context)) {
+					if (!(usage in translations)) {
+						translations[usage] = []
+					}
+					// split the translations by usage in components
+					translations[usage].push({
+						l: locale,
+						// eslint-disable-next-line @typescript-eslint/no-unused-vars
+						t: Object.fromEntries(Object.entries(currentTranslations).filter(([id, _value]) => msgIds.includes(id))),
+					})
+				}
+			}
+		},
+
+		/**
+		 * Hook into module resolver and fake all '../[...]/l10n.js' imports to inject our splitted translations
+		 * @param source The file which is imported
+		 * @param importer The file that imported the file
+		 */
+		resolveId(source, importer) {
+			if (source.startsWith('\0')) {
+				if (source === '\0l10n') {
+					// return our l10n main module containing all translations
+					return '\0l10n'
+				}
+				// dont handle other plugins imports
+				return false
```

Comment:
Are you sure that we should ignore others here and not return to the same source for further resolving? 

```suggestion
				// dont handle other plugins imports
				return source
```

Reply:
No, but false is also wrong. I mixed it up, from the docs:

> Returning null defers to other resolveId functions and eventually the default resolution behavior.
> Returning `false` signals that source should be treated as an external module and not included in the bundle.

Reply:
fixed :heavy_check_mark: